src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -795,203 +789,187 @@ where
+            storage_metadata.clone(),
+            now.clone(),
+            timestamp_interval,
+            as_of,
+        )
+        .await
+        {
+            Ok(t) => t,
+            Err(e) => {
+                panic!("Failed to create source {} timestamper: {:#}", name, e);
             }
+        };
+        // The global view of the source_upper, which we track by combining
+        // summaries from the raw reader operators.
+        let mut global_source_upper = timestamper
+            .source_upper_at_frontier(resume_upper.borrow())
+            .expect("source_upper_at_frontier to be used correctly");
+
+        let new_ts_upper = timestamper
+            .reclock_frontier(&global_source_upper)
+            .expect("compacted past upper");
+
+        cap_set.downgrade(new_ts_upper);
+
+        // Emit initial snapshot of the remap_shard, bootstrapping
+        // downstream reclock operators.
+        let remap_trace = timestamper.remap_trace();
+        trace!(
+            "remap({id}) {worker_id}/{worker_count}: \
+                emitting initial remap_trace. \
+                source_upper: {:?} \
+                trace_updates: {:?}",
+            global_source_upper,
+            remap_trace
+        );
+
+        // Out of an abundance of caution, do not hold the output handle
+        // across an await, and drop it before we downgrade the capability.
+        {
+            let mut remap_output = remap_output.activate();
+            let cap = cap_set.delayed(cap_set.first().unwrap());
+            let mut session = remap_output.session(&cap);
+            session.give(remap_trace);
+        }
 
-            // The last frontier we compacted the remap shard to, starting at [0].
-            let mut last_compaction_since = Antichain::from_elem(Timestamp::default());
-
-            while scheduler.notified().await {
-                if token_weak.upgrade().is_none() {
-                    // Make sure we don't accidentally mint new updates when
-                    // this source has been dropped. This way, we also make sure
-                    // to not react to spurious frontier advancements to `[]`
-                    // that happen when the input source operator is shutting
-                    // down.
-                    return;
-                }
+        // The last frontier we compacted the remap shard to, starting at [0].
+        let mut last_compaction_since = Antichain::from_elem(Timestamp::default());
 
-                if !active_worker {
-                    // We simply loop because we cannot return here. Otherwise
-                    // the one active worker would not get frontier upgrades
-                    // anymore or other things could break.
+        tokio::pin!(shutdown_rx);
+        let mut input_frontier = Antichain::from_elem(Timestamp::default());
+        loop {
+            // AsyncInputHandle::next is cancel safe
+            tokio::select! {
+                biased;
```

Comment:
can you add a comment to switch this to (as yet unwritten) `select3`? its also biased, so it would drop in perfectly